### PR TITLE
Lightning

### PR DESCRIPTION
**Pull request by amrukwa**

### CODE REVIEW

File: semantic_segmentation/scripts/train.py [hunk 1]
```diff
@@ -1,23 +1,78 @@
...
-from torch.utils.tensorboard import SummaryWriter
 from torchmetrics import Accuracy
-from tqdm import tqdm
 from unet_bn import U2NET_lite
 
+from augment import get_transform
+from dataset import Dataset
+
+
+class UNET_LT(pl.LightningModule):
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: Let's change the name to something not related to the architecture
Something challenge related?

In the future we will change the net

File: semantic_segmentation/scripts/train.py [hunk 2]
```diff
@@ -1,23 +1,78 @@
...
+    def __init__(self, lr, **kwargs):
+        super().__init__()
+        self.Net = U2NET_lite()
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: Use lowercase for the net

File: semantic_segmentation/scripts/train.py [hunk 3]
```diff
@@ -1,23 +1,78 @@
...
+        self.Net.cuda(1)
+        self.loss = nn.CrossEntropyLoss(ignore_index=255)
+        self.lr = lr
+        self.accuracy = Accuracy(
+            task="multiclass", num_classes=16, ignore_index=255
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: Let's have this 16 parametrized
And pass it also to the u2net constructor

We will need such a param when we will change arch

File: semantic_segmentation/scripts/train.py (hunk 1 above)
Comment (amrukwa) [reply semantic_segmentation/scripts/train.py]: sure

File: semantic_segmentation/scripts/train.py [hunk 4]
```diff
@@ -1,23 +1,78 @@
...
+        ).cuda(1)
+        self.save_hyperparameters()
+
+    @staticmethod
+    def add_model_specific_args(parent_parser):
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py:30]: Nice!

File: semantic_segmentation/scripts/train.py [hunk 5]
```diff
@@ -1,23 +1,78 @@
...
+        parser = parent_parser.add_argument_group("U2NetModel")
+        parser.add_argument("--lr", type=float, default=3e-4)
+        return parent_parser
+
+    def forward(self, x):
+        pred = self.Net(x)
+        return pred
+
+    def configure_optimizers(self):
+        optimizer = optim.RAdam(self.parameters(), lr=self.lr)
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py:41]: Later - here we'll add scheduler

File: semantic_segmentation/scripts/train.py [hunk 6]
```diff
@@ -1,23 +1,78 @@
...
+        return optimizer
+
+    def training_step(self, train_batch, batch_idx):
+        img, target = train_batch
+        img = img.cuda(1)
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: Remove explicit .cuda
Lighting automatically copies the data - your batch should be already on gpu

If you want to use gpu1 instead of first check Trainer arg - there you can set how many gpus or just pass the gpu ids

File: semantic_segmentation/scripts/train.py [hunk 7]
```diff
@@ -1,23 +1,78 @@
...
+        target = target.cuda(1)
+        pred = self.Net(img)
+        loss = self.loss(pred, torch.squeeze(target.long()))
+        acc = self.accuracy(torch.argmax(pred, dim=1), torch.squeeze(target))
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py:48]: This one without grad (e.g. pred detach)

File: semantic_segmentation/scripts/train.py [hunk 8]
```diff
@@ -1,23 +1,78 @@
...
+        self.log("train_loss", loss)
+        self.log("train_acc", acc)
+        if batch_idx == 0:
+            tensorboard = self.logger.experiment
+            tensorboard.add_figure("predictions", show_results(img, pred, target))
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: Do you need the tensorboard var?

File: semantic_segmentation/scripts/train.py [hunk 9]
```diff
@@ -1,23 +1,78 @@
...
+        return loss
+
+    def validation_step(self, val_batch, batch_idx):
+        img, target = val_batch
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: If you think about it you can merge train and val steps into step
With additional param - split (train or val as str)

**Review by karolmajek — CHANGES_REQUESTED** [review]

File: semantic_segmentation/scripts/train.py (hunk 8 above)
Comment (amrukwa) [reply semantic_segmentation/scripts/train.py]: yes, self.log only operates on scalars, as per docs

Comment (karolmajek) [reply semantic_segmentation/scripts/train.py]: this:
`self.logger.experiment.add_figure("predictions", show_results(img, pred, target))`

File: semantic_segmentation/scripts/train.py [hunk 10]
```diff
@@ -20,16 +20,15 @@
 from dataset import Dataset
 
 
-class UNET_LT(pl.LightningModule):
-    def __init__(self, lr, **kwargs):
+class Suad_Semseg(pl.LightningModule):
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: `Class names should normally use the CapWords convention.`
https://peps.python.org/pep-0008/#class-names

File: semantic_segmentation/scripts/train.py [hunk 11]
```diff
@@ -20,16 +20,15 @@
...
+    def __init__(self, arch, lr, num_classes, **kwargs):
         super().__init__()
-        self.Net = U2NET_lite()
-        self.Net.cuda(1)
+        self.net = arch()
```
Comment (karolmajek) [inline semantic_segmentation/scripts/train.py]: after a while - it will be better to pass the object instance named as `net`:

```
self.net = net
```

more intuitive

**Review by karolmajek — CHANGES_REQUESTED** [review]

**Comment by karolmajek** [comment]
just run it to be sure, before merging

**Review by karolmajek — APPROVED** [review]
